File: src/main/java/com/sparkbook/java/BasicAvg.java
package com.sparkbook.java;

import org.apache.spark.api.java.JavaDoubleRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.DoubleFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.storage.StorageLevel;


import java.io.Serializable;
import java.util.Arrays;

public final class BasicAvg {
    public static class AvgCount implements Serializable {
        public int total;
        public int num;

        public AvgCount(int total, int sum) {
            this.total = total;
            this.num = sum;
        }
        public float avg() {
            return total/num;
        }
    }

    public static void main(String[] args) {
        String master;
        if (args.length > 0) {
            master = args[0];
        } else {
            master = "local";
        }

        JavaSparkContext sc = new JavaSparkContext(master, "basicavg", System.getenv("SPARK_HOME"),System.getenv("JARS"));
        JavaRDD<Integer> rdd = sc.parallelize(Arrays.asList(1,2,3,4));
        JavaDoubleRDD result = rdd.mapToDouble(
                new DoubleFunction<Integer>() {
                    @Override
                    public double call(Integer x) throws Exception {
                        return (double) x*x;
                    }
                }
        );
//        result.persist(StorageLevel.DISK_ONLY());
//        System.out.println(result.mean());
//        Function2<AvgCount, Integer, AvgCount> addAndCount = new Function2<AvgCount, Integer, AvgCount>() {
//
//            public AvgCount call(AvgCount a, Integer b) throws Exception {
//                a.total += b;
//                a.num += 1;
//                return a;
//            }
//        };
//
//        Function2<AvgCount, AvgCount, AvgCount> combine = new Function2<AvgCount, AvgCount, AvgCount>() {
//            public AvgCount call(AvgCount a, AvgCount b) throws Exception {
//                a.total += b.total;
//                a.num += b.num;
//                return a;
//            }
//        };
//        AvgCount initial = new AvgCount(0,0);
//        AvgCount result = rdd.aggregate(initial,addAndCount,combine);
//        System.out.println(result.avg());
//        System.out.println(rdd.countByValue());
//        sc.stop();
    }
}

File: src/main/scala/com/sparkbook/scala/BasicAvg.scala
package com.sparkbook.scala

import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD

object BasicAvg {

  def computeAvg(input: RDD[Int]) = {
    input.aggregate((0,0))((x,y) => (x._1+y, x._2 + 1),
      (x,y) => (x._1 + y._1, x._2 + y._2))
  }


  def main(args: Array[String]): Unit = {
    val master = args.length match {
      case x: Int if x > 0 => args(0)
      case _ => "local"
    }

    val sc = new SparkContext(master,"BasicAvg",System.getenv("SPARK_HOME"))
    val input = sc.parallelize(List(1,2,3,4))
    val result = computeAvg(input)
    val avg = result._1 / result._2.toFloat
    println("lihe",avg)
  }
}
